### PR TITLE
Update HistogramFn impl to log failure to record rather than panic

### PR DESCRIPTION
**Pull request by dannycjones**
## Description of change

When writing 50GiB of data to a file, mountpoint-s3 panicked due to an unexpected "ValueOutOfRangeResizeDisabled" error. When creating the Histogram, we limit it with an upper bound of 300000000. For time durations in microsecond precision, that's 300 seconds. In this case, the file was being written to for around 15 minutes (or 900 seconds).

I'm not sure why we landed on 300 seconds for the upper limit on the histogram, but for now this will ensure mountpoint-s3 does not crash when it cannot record a histogram value but instead continue with a loud `ERROR` message in the logs.

After this change:

```
2023-09-11T16:46:51.130810Z ERROR open{req=12 ino=2}:put_object{id=7 bucket="***" key="50GiB_001.bin"}: mountpoint_s3::metrics::data: metric value 1078980958.0 could not be recorded: ValueOutOfRangeResizeDisabled
2023-09-11T16:46:51.131035Z ERROR open{req=12 ino=2}:put_object{id=7 bucket="***" key="50GiB_001.bin"}: mountpoint_s3::metrics::data: metric value 1078980958.0 could not be recorded: ValueOutOfRangeResizeDisabled
2023-09-11T16:46:51.133947Z ERROR mountpoint_s3::metrics::data: metric value 1079032281.0 could not be recorded: ValueOutOfRangeResizeDisabled
```

## Does this change impact existing behavior?

No breaking changes, just allows mountpoint-s3 to continue operating after receiving a histogram value too large.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3/src/metrics/data.rs [hunk 1]
```diff
@@ -147,15 +149,24 @@ impl AtomicGauge {
...
 impl metrics::HistogramFn for Histogram {
     fn record(&self, value: f64) {
-        self.histogram.lock().unwrap().record(value as u64).unwrap();
+        let mut histogram = self.histogram.lock().unwrap();
+        match histogram.record(value as u64) {
+            Ok(_) => {}
+            Err(e) => {
+                error!("metric value {:?} could not be recorded: {:?}", value, e);
+            }
+        }
```
Comment (jamesbornholt) [inline mountpoint-s3/src/metrics/data.rs:169]: Think we should probably just use `saturating_record` here instead -- 300s is long enough that we can pretty much interpret values in that bucket as "saturated".
```suggestion
        self.histogram.lock().unwrap().saturating_record(value as u64);
```

**Review by jamesbornholt — COMMENTED** [review]
The 300s number is an arbitrary one I picked to constrain memory usage. It looks like `hdrhistogram` supports auto-resizing, so maybe we should just use that instead?

File: mountpoint-s3/src/metrics/data.rs (hunk 1 above)
Comment (dannycjones) [reply mountpoint-s3/src/metrics/data.rs:169]: Makes sense, although I'm a little worried that this is difficult to know as a customer. I'll see if there's any documentation that I can add a note to.

**Comment by dannycjones** [comment]
Ah, I saw the autoresizing but figured there may be some downsides. I'll take a deeper look. If its fine, then we can avoid the unexpected clamping at 300s.

File: mountpoint-s3/src/metrics/data.rs (hunk 1 above)
Comment (dannycjones) [reply mountpoint-s3/src/metrics/data.rs]: I've instead enabled resizing and moved back to `record` with an `expect`.

File: mountpoint-s3/src/metrics/data.rs [hunk 2]
```diff
@@ -1,6 +1,8 @@
 use crate::sync::atomic::{AtomicU64, AtomicUsize, Ordering};
 use crate::sync::{Arc, Mutex};
 
+use tracing::error;
```
Comment (dannycjones) [inline mountpoint-s3/src/metrics/data.rs:4]: and need to drop this :)

**Review by jamesbornholt — APPROVED** [review]

**Pull request merged**